resource/Resource.md

```diff
@@ -0,0 +1,133 @@
+quoted and `"` characters MUST be escaped.
+
+Population from environment variables MUST be the first applied detection process unless
+the user explicitly overwrites this behavior.
+
+### Auto-detection
+Auto-detection of resource information in specific environments, e.g. specific cloud
+vendors, MUST be implemented outside of the core libraries in third party or
+[census-ecosystem][census-ecosystem] repositories.
+
+### Merging
+When new information is gained about a resource, already set labels or type fields MUST NOT
+be overwritten. Label key namespaceing SHOULD be used to prevent collisions across different
+resource detection steps.
+
+### Detectors
+To make auto-detection implementations easy to use, the core resource package SHOULD define
+an interface to retrieve resource information. Additionally, helper functionality MAY be
+provided to effectively make use of this interface.
+The exact shape of those interfaces and helpers SHOULD be idiomatic to the respective language.
+
+Example in Go:
+
+```go
+type Detector func(context.Context) (*Resource, error)
+
+// Returns a detector that always returns a specific resource.
+func NewDetectorFromResource(*Resource) Detector
+
+// Returns a detector that runs all input detectors sequentially and merges their results.
+func ChainedDetector(...Detector) Detector
+```
+
+## Exporter translation
+A resource object MUST NOT be mutated further once it is passed to a backend-specific exporter.
+From the provided resource information, the exporter MAY transform, drop, or add information
+to build the resource identifying data type specific to its backend.
+If the passed resource does not contain sufficient information, an exporter MAY drop
+signal data entirely, if no sufficient resource information is provided to perform a correct
+write.
+
+For example, from a resource object
+
+```javascript
+{
+	"type": "kubernetes.io/container",
+	"labels": {
+		// Populated from VM environment through auto-detection library.
+		"cloud.google.com/gce/instance_id": "instance1",
+		"cloud.google.com/zone": "eu-west2-a",
+		"cloud.google.com/project_id": "project1",
+		"cloud.google.com/gce/attributes/cluster_name": "cluster1",
+		// Populated through OpenCensus resource environment varibales.
+		"kubernetes.io/namespace": "ns1",
+		"kubernetes.io/pod_name": "pod1",
+		"kubernetes.io/container_name": "container1",
+	},
+}
+```
+
+an exporter for Stackdriver would create the following "monitored resource", which is a
+resource type with well-known identifiers specific to its API:
+
+```javascript
+{
+	"type": "k8s_container",
+	"labels": {
+		"project_id": "project1",
+		"location": "eu-west2-a",
+		"cluster_name": "cluster1",
+		"namespace_name": "ns1",
+		"pod_name": "pod1",
+		"container_name": "container1",
+	},
+}
+```
+
+Exporter libraries MAY provide a default translation for well-known input resource types and labels.
```

Comment:
It might be good to add a list of all known label keys here in the future.

Then, implementors can look up for the whole list and decide what they can map.